### PR TITLE
[improvement](page builder) avoid allocating big memory in ctor

### PR DESCRIPTION
**Pull request by jacktengg**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by jacktengg** [comment]
run buildall

File: be/src/olap/rowset/segment_v2/binary_dict_page.cpp [hunk 1]
```diff
@@ -45,17 +46,27 @@ BinaryDictPageBuilder::BinaryDictPageBuilder(const PageBuilderOptions& options)
           _finished(false),
           _data_page_builder(nullptr),
           _dict_builder(nullptr),
-          _encoding_type(DICT_ENCODING) {
+          _encoding_type(DICT_ENCODING) {}
+
+Status BinaryDictPageBuilder::init() {
+    RETURN_IF_ERROR_OR_CATCH_EXCEPTION(_init());
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/binary_dict_page.cpp:52]: add a flag to indicate init is called
and add dcheck to all method to check other method is called after init.

**Comment by jacktengg** [comment]
run buildall

File: be/src/olap/rowset/segment_v2/page_builder.h [hunk 2]
```diff
@@ -91,5 +96,14 @@ class PageBuilder {
     DISALLOW_COPY_AND_ASSIGN(PageBuilder);
 };
 
+template <typename Derived>
+class PageBuilderHelper : public PageBuilder {
+public:
+    template <typename... Args>
+    static Derived* create(Args&&... args) {
+        return new Derived(std::forward<Args>(args)...);
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/page_builder.h:104]: why not call init here

Comment (jacktengg) [reply be/src/olap/rowset/segment_v2/page_builder.h:104]: 在改了

File: be/src/olap/rowset/segment_v2/page_builder.h [hunk 3]
```diff
@@ -91,5 +96,14 @@ class PageBuilder {
     DISALLOW_COPY_AND_ASSIGN(PageBuilder);
 };
 
+template <typename Derived>
+class PageBuilderHelper : public PageBuilder {
+public:
+    template <typename... Args>
+    static Derived* create(Args&&... args) {
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/page_builder.h:103]: static Status create(Derived*, Args&&... args)

File: be/src/olap/rowset/segment_v2/page_builder.h [hunk 4]
```diff
@@ -91,5 +94,15 @@ class PageBuilder {
     DISALLOW_COPY_AND_ASSIGN(PageBuilder);
 };
 
+template <typename Derived>
+class PageBuilderHelper : public PageBuilder {
+public:
+    template <typename... Args>
+    static Status create(PageBuilder** builder, Args&&... args) {
+        *builder = new Derived(std::forward<Args>(args)...);
```
Comment (yiguolei) [inline be/src/olap/rowset/segment_v2/page_builder.h:102]: use a unique ptr
if init failed, then it is auto released.
if status is error, user may not release the builder manually.

**Comment by jacktengg** [comment]
run buildall

**Review by yiguolei — APPROVED** [review]

**Review by zclllyybb — APPROVED** [review]

**Review by Gabriel39 — APPROVED** [review]

**Pull request merged**